### ISSUE TITLE
[$250] Workspace - `That page is nowhere to be found` error message appears for a brief moment reported by @thesahindia 

### ISSUE THREAD

**Issue by kavimuru** [Expensify/App#12616]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

## Action Performed:

1. Navigate to settings > Workspaces
2. Select a workspace
3. Copy the URL
4. Sign out > paste the URL and press enter
5. Now login

## Expected Result:  
You should see a loading indicator

## Actual Result: 
Not found view is being displayed

## Workaround:
unknown

## Platform:

Where is this issue occurring?

- Web
- Desktop App
- Mobile Web

**Version Number:** 1.2.25-0
**Reproducible in staging?:** y
**Reproducible in production?:** y
**Email or phone of affected tester (no customers):** 
**Logs:** https://stackoverflow.com/c/expensify/questions/4856
**Notes/Photos/Videos:** 

https://user-images.githubusercontent.com/43996225/200940349-735abcc7-110d-4bf4-a3c8-d99a0b518b43.mov

https://user-images.githubusercontent.com/43996225/200940364-a9196aa3-b943-4d05-a2b4-80893ead3a61.mp4

**Expensify/Expensify Issue URL:** 
**Issue reported by:** @thesahindia 
**Slack conversation:** https://expensify.slack.com/archives/C01GTK53T8Q/p1667978507958729

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

**Comment by laurenreidexpensify** [comment]
https://www.upwork.com/jobs/~01d19e142ce5b9a8ec

**Comment by ma3ve** [comment]
Proposal: 
The issue might be because the error message defaults when you are not authenticated. so until the authentication data is fetched, we will be shown the error message. so to solve this, we need to add a conditional check for if the authentication data is being fetched. if it is being fetched then don't show the error.

**Comment by s77rt** [comment]
### Proposal
```diff
diff --git a/src/pages/workspace/WorkspaceInitialPage.js b/src/pages/workspace/WorkspaceInitialPage.js
index 1c016c601..f7e5cfdb8 100644
--- a/src/pages/workspace/WorkspaceInitialPage.js
+++ b/src/pages/workspace/WorkspaceInitialPage.js
@@ -148,7 +148,7 @@ class WorkspaceInitialPage extends React.Component {
 
         return (
             <ScreenWrapper>
-                <FullPageNotFoundView shouldShow={_.isEmpty(this.props.policy)}>
+                <FullPageNotFoundView shouldShow={this.props.isLoadingInitialAppData === false && _.isEmpty(this.props.policy)}>
                     <HeaderWithCloseButton
                         title={this.props.translate('workspace.common.workspace')}
                         shouldShowBackButton
@@ -274,5 +274,8 @@ export default compose(
         reports: {
             key: ONYXKEYS.COLLECTION.REPORT,
         },
+        isLoadingInitialAppData: {
+            key: ONYXKEYS.IS_LOADING_INITIAL_APP_DATA,
+        },
     }),
 )(WorkspaceInitialPage);
```

### Details
The fix is make sure the app is loaded before displaying the `FullPageNotFoundView` so we can act based on the real collected data.
PS: The fix should be applied on other pages as well such as WorkspaceSettingsPage, WorkspaceInvitePage, etc.

#### Plus:
With the proposed fix, the name of the workspace will appear empty, then it will be populated. We can fix that by moving the isLoadingInitialAppData check a little above to return null or a spinner.
```diff
diff --git a/src/pages/workspace/WorkspaceInitialPage.js b/src/pages/workspace/WorkspaceInitialPage.js
index 1c016c601..ce9e8e474 100644
--- a/src/pages/workspace/WorkspaceInitialPage.js
+++ b/src/pages/workspace/WorkspaceInitialPage.js
@@ -95,6 +95,10 @@ class WorkspaceInitialPage extends React.Component {
     }
 
     render() {
+        if (this.props.isLoadingInitialAppData !== false) {
+            return null;
+        }
+
         const policy = this.props.policy;
         const hasMembersError = PolicyUtils.hasPolicyMemberError(this.props.policyMemberList);
         const hasGeneralSettingsError = !_.isEmpty(lodashGet(this.props.policy, 'errorFields.generalSettings', {}))
@@ -274,5 +278,8 @@ export default compose(
         reports: {
             key: ONYXKEYS.COLLECTION.REPORT,
         },
+        isLoadingInitialAppData: {
+            key: ONYXKEYS.IS_LOADING_INITIAL_APP_DATA,
+        },
     }),
 )(WorkspaceInitialPage);
```


PS: This fix is a per page fix, means it must be done on all pages that should not render before init. To make the global fix, we can port the isLoadingInitialAppData check (or better a promise) to the modal stack level (the sidebar).


https://user-images.githubusercontent.com/16493223/201131875-a6108874-e924-4d22-8a7d-087d29d9598e.mp4

**Comment by mananjadhav** [comment]
Thanks for the proposal @s77rt.  I agree with your analysis.

I am wondering if this check should be at `FullPageNotFoundView` page itself? What do you think @Beamanator? That way it is fixed across all the pages. We can fallback to show a loader when `isLoadingInitialAppData` is true?

**Comment by Beamanator** [comment]
@s77rt Thanks for the analysis!

@mananjadhav I like this idea as well, I think it makes sense to apply this "fix" across all pages using `FullPageNotFoundView` - I want to get @Expensify/design 's opinion as well, though it looks like @shawnborton is OOO till Nov 17th

@laurenreidexpensify would you mind double-checking the proposed solution? Basically what you can see in the videos in [this post](https://github.com/Expensify/App/issues/12616#issuecomment-1310436422) is that the Workspace page briefly shows as blank while loading the authenticated information, instead of briefly showing `That page is nowhere to be found` (see video in OP)

**Comment by laurenreidexpensify** [comment]
The blank page sounds reasonable, but does that just hide the underlying issue instead of directly addressing it?

**Comment by mananjadhav** [comment]
I was going to recommend a spinner instead of a blank page and it is only going to be temporary.

**Comment by daraksha-dk** [comment]
**Proposal**

To display a loading indicator a I propose these changes

```diff
 const FullPageNotFoundView = (props) => {
+    if (props.isLoadingInitialAppData) {
+        return <FullScreenLoadingIndicator />;
+    }
     if (props.shouldShow) {
         return (
             <>
@@ -74,4 +81,11 @@ FullPageNotFoundView.propTypes = propTypes;
 FullPageNotFoundView.defaultProps = defaultProps;
 FullPageNotFoundView.displayName = 'FullPageNotFoundView';
 
-export default withLocalize(FullPageNotFoundView);
+export default compose(
+    withOnyx({
+        isLoadingInitialAppData: {
+            key: ONYXKEYS.IS_LOADING_INITIAL_APP_DATA,
+        },
+    }),
+    withLocalize,
+)(FullPageNotFoundView);
```
at https://github.com/Expensify/App/blob/main/src/components/BlockingViews/FullPageNotFoundView.js

**Comment by Beamanator** [comment]
@mananjadhav I think it would be nice to see what a spinner would look like before agreeing it's a better solution - you 100% could be right but let's see first :D Maybe @daraksha-dk can show what their proposal would look like with a video?

@laurenreidexpensify I think there's no real "underlying issue" here that we're "avoiding fixing" - it's just a situation where the app is logging in and immediately redirecting to the workspace settings (or other pages) whileee we're fetching some information about the workspace

**Comment by s77rt** [comment]
@Beamanator 
I think @daraksha-dk proposal's is not substantively different from mine, is it?

**Comment by daraksha-dk** [comment]
> Maybe @daraksha-dk can show what their proposal would look like with a video?

This is how it will look like

https://user-images.githubusercontent.com/71884990/201409359-ab512d16-153c-493d-8716-e862a2b86e33.mov

**Comment by JmillsExpensify** [comment]
@laurenreidexpensify My two cents is that the spinner is superior. I think the app looks broken even with a blank page.

**Comment by Beamanator** [comment]
I think I agree, it is quite nice to have _something_ there - what about a skeleton view thing instead of a spinner? 🤔 I

**Comment by JmillsExpensify** [comment]
Now we're talking! That's a super great idea. It's easy to default to spinner, but even that can be improved upon.

**Comment by Beamanator** [comment]
Let's see if we can get a video of what this could look! Too many spinners drives me a bit crazy 😅

**Comment by laurenreidexpensify** [comment]
Nice I like the skeleton too. Do we need a mock up from @shawnborton here for that?

**Comment by Beamanator** [comment]
Ooh yes I think so @laurenreidexpensify 👍 Looks like Shawn comes back from OOO today, so maybe we can give him a few days 😅

**Comment by shawnborton** [comment]
I don't particularly love having two loading spinners on the page like that. What if we just had the general loading spinner for the site and then once everything is loaded, we open up the modal?

**Comment by Beamanator** [comment]
Just so we're on the same page @shawnborton - what about showing the Skeleton UI in the modal while it's loading (& keep the main spinner)?

I think I could also get behind your idea of showing 1 general loading spinner, then opening the modal once everything is loaded

**Comment by JmillsExpensify** [comment]
Oh just commented in a related issue. I generally like showing the skeleton UI in the modal while loading, and I guess perhaps more generally, in what scenario(s) do we prefer the spinner vs skeleton UI. We should probably align on the broader pattern at play.

**Comment by Beamanator** [comment]
> We should probably align on the broader pattern at play.
100%

**Comment by mananjadhav** [comment]
I agree with the idea of adding a skeleton UI compared to a common/multiple spinners.

**Comment by shawnborton** [comment]
Hmm but the skeleton UI is meant to look like a list of chats or a group of messages. Why would we use that design to load a page that is not a group of chats or messages?

**Comment by Beamanator** [comment]
@shawnborton aah interesting, I assumed the skeleton UI could be used in other pages where we know the shape of the page that's being loaded - i guess skeleton could be overkill here since we already know 99% of what will show up in the Workspace settings page, the only thing we don't immediately know is the workspace name & avatar (unless they're cached in Onyx).

Is it better to instantly load the page (without waiting for onyx data), with default avatar & a blank space where the name will go? Or just don't open the page until all the data is loaded and ready to be displayed?

**Comment by shawnborton** [comment]
How would we load this page if the user was 100% offline?

**Comment by Beamanator** [comment]
Sorry what exact flow are you asking about @shawnborton ?

**Comment by shawnborton** [comment]
The issue that we are talking about here, where we are recommending using a skeleton UI. That feels overkill, so I was curious what we do when we load the workspace editor when the user is offline.

**Comment by Beamanator** [comment]
Hmm well I think the issue being tackled here is only this case:
1. User tries navigating to a URL that's some modal page, but they weren't logged in
2. User is redirected to the sign in page, then they sign in
3. Once they sign in, they're redirected to the original URL, and Onyx data starts loading

So if the user is offline (but signed in), opening the workspace settings / editor should just open instantly since Onyx data was already loaded into the app, right?

**Comment by Beamanator** [comment]
> For this particular scenario where we are just loading a workspace name/avatar, how long do we think that's going to take to load each time? If it should be fairly brief, skeleton feels like overkill.

I think this is a good point, it definitely looks like the loading time is very brief - maaaybe a bit longer if the user is on very slow network connection 🤷 

I like the idea of making a mini skeleton UI just for avatar & name of the workspace, OR trying to make the modal not open at all until the data in Onyx is ready to be rendered. Also interested to hear other people's opinions on this :D

**Comment by JmillsExpensify** [comment]
Jumping back in, awesome convo! Hmm, ok I see where Shawn is coming from. In fact, while the issue we're talking about here is different given that we're talking about web/desktop, I would love to avoid the grey/flashing we've seen elsewhere. So I think I'd back to either:

 - Showing the spinner for the workspace modal
 - Not opening the modal until all Onyx data is loaded. 

I think I'm honestly back to spinner so that the general page can load fast, even if the workspace modal takes longer.

**Comment by daraksha-dk** [comment]
The problem with implementing skeleton UI is that it needs to be added for every page where this issue is reproducible. All the workspace pages are different so we need different skeleton UI components that matches to the specific page.

> Showing the spinner for the workspace modal

I prefer this because it's simple to implement

> Not opening the modal until all Onyx data is loaded.

I think this can be complex and it can cause regressions too. Because this issue is an edge case I do not think it's worth it.

**Comment by shawnborton** [comment]
Cool, I can get down with the spinner approach as well.

**Comment by mananjadhav** [comment]
I also feel spinner is the right approach. Because we are adding it to a generic `NotFound` page.

**Comment by JmillsExpensify** [comment]
Awesome discussion all around. Let's move forward with spinner!

**Comment by mananjadhav** [comment]
@daraksha-dk's proposal [here](https://github.com/Expensify/App/issues/12616#issuecomment-1311654012) looks good to go ahead.


🎀 👀 🎀 
C+ reviewed

**Comment by s77rt** [comment]
@mananjadhav That's basically my proposal. It is not substantively different than mine.

**Comment by mananjadhav** [comment]
@s77rt it isn't but it was complete with spinner instead of the blank page, covers the fact that it can be a part of NotFound page. Lastly @daraksha-dk was actively involved in the last conversations, giving their insights, feedback, etc.

Hope it clarifies the decision.

**Comment by s77rt** [comment]
@mananjadhav I did mention to use a spinner, the proposal does not have to be on the final version. And the fact to use on the NotFound page was suggested by @Beamanator

**Comment by mananjadhav** [comment]
Thanks for the comment @s77rt. I didn't just mention spinner, but few other things. I'll let @Beamanator share their views on this one.

**Comment by laurenreidexpensify** [comment]
Bump @Beamanator ^^

**Comment by Beamanator** [comment]
Cool just catching up here! It looks like we're going double-spinner! 👍  It will be very quick anyway, so I'm down for that solution 👍 

As for hiring, I agree with @mananjadhav it makes sense to go with @daraksha-dk since their proposal + solution was exactly what we ended up going with (spinners) & I don't think it was "too similar" to the other proposal since the accepted proposal was concise (updating 1 component instead of many pages)

**Comment by thesahindia** [comment]
Looks like it's not an issue the user shouldn't be able to visit the page like this. C/P Rajat's comment from the [thread](https://expensify.slack.com/archives/C02NK2DQWUX/p1669114217756719?thread_ts=1669075338.989349&cid=C02NK2DQWUX) -

"There are several reports like this one from [@0xmiroslav](https://github.com/0xmiroslav). They all are related to a bug in RN, I have requested to hold 3 of them.
They are incorrectly titled as deeplink but this behavior does not come under deeplinks. we should hold working on any such issues unless we want to app to have this behaviour. In that case, this should be title as feature request.
App was never configured to handle these cases and we never expected  for someone to be redirected to pages like this. There is a bug in RN which is causing this navigational behavior.
This RN bug is now being tracked with main navigation Root tracking issue."

**Comment by thesahindia** [comment]
@Beamanator @mananjadhav let's hold this issue until the next steps are clear.

**Comment by laurenreidexpensify** [comment]
Not overdue- waiting for next step confirmation from @mananjadhav and @Beamanator

**Comment by Beamanator** [comment]
@thesahindia thanks for noting that this issue is related to a few others, and it seems we haven't properly discussed how we want to fix it! 

Sorry @daraksha-dk I'm unassigning you for now, since I think we should hold this issue as well until we come up with a better solution for the entire app 👍 

@thesahindia / @parasharrajat is there a "bigger solution" discussion happening in one of the related issues?

**Comment by parasharrajat** [comment]
@Beamanator these are my thoughts https://expensify.slack.com/archives/C02NK2DQWUX/p1669114217756719?thread_ts=1669075338.989349&cid=C02NK2DQWUX on it.

**Comment by Beamanator** [comment]
@parasharrajat can you link the tracking issue here?

**Comment by parasharrajat** [comment]
On https://github.com/Expensify/App/issues/11768 under **Drawer Navigation / State** title.

**Comment by JmillsExpensify** [comment]
Based on the more recent updates, I'm going to put this issue on hold and link it to #11768.

**Comment by JmillsExpensify** [comment]
Related, given that I'm the BZ assigned to the tracking issue, I'm assigning myself and unassigning Lauren.

**Comment by Beamanator** [comment]
Thanks @parasharrajat & thanks @JmillsExpensify for adding this issue to the tracker - since this is on hold I'm going to drop it to `Weekly` for now

**Comment by JmillsExpensify** [comment]
Great call, I missed that one.

**Comment by JmillsExpensify** [comment]
Still on hold for the larger linked initiative.

**Comment by JmillsExpensify** [comment]
Still on hold for the larger react navigation re-write.

**Comment by JmillsExpensify** [comment]
Still on hold for the navigation re-write.

**Comment by JmillsExpensify** [comment]
Still on hold for navigation revamp.

**Comment by JmillsExpensify** [comment]
Switched out weekly for monthly given that this is a larger project with no defined ETA.

**Comment by JmillsExpensify** [comment]
Still on hold for the navigation project.

**Comment by JmillsExpensify** [comment]
Same same.

**Comment by JmillsExpensify** [comment]
Still on hold.

**Comment by Beamanator** [comment]
Seems like we're still held but there's lots of [progress](https://github.com/Expensify/App/issues/11768#issuecomment-1518477652)!

**Comment by Beamanator** [comment]
still on hollld

**Comment by Beamanator** [comment]
Ooh navigation project is done, right?? Maybe this can be off hold finally?

**Comment by luacmartins** [comment]
what's the latest here @Beamanator? are we closing this issue and creating a new one?

**Comment by Beamanator** [comment]
I just tried to reproduce & now I see a loader on the workspace RHN so I think we can close this out
